internal/ted448/scalar.go

```diff
@@ -162,42 +184,68 @@ func (z *Scalar) Add(x, y *Scalar) {
 	y64.fromScalar(y)
 	c := add(z64[:], x64[:], y64[:])
 	add(t[:], z64[:], residue448[:])
-	z64.Cmov(c, &t)
+	z64.cmov(&t, c)
 	z64.modOrder()
 	z64.toScalar(z)
 }
 
 // Sub calculates z = x-y mod order.
 func (z *Scalar) Sub(x, y *Scalar) {
-	var z64, x64, y64, t scalar64
+	var z64, x64, y64 scalar64
 	x64.fromScalar(x)
 	y64.fromScalar(y)
-	c := sub(z64[:], x64[:], y64[:])
-	sub(t[:], z64[:], residue448[:])
-	z64.Cmov(c, &t)
-	z64.modOrder()
+	z64.sub(&x64, &y64)
 	z64.toScalar(z)
 }
 
 // Mul calculates z = x*y mod order.
 func (z *Scalar) Mul(x, y *Scalar) {
 	var z64, x64, y64 scalar64
-	prod := (&[_N + 1]uint64{})[:]
 	x64.fromScalar(x)
 	y64.fromScalar(y)
+	coremul(&z64, &x64, &y64)
+	z64.modOrder()
+	z64.toScalar(z)
+}
+
+func coremul(z64, x64, y64 *scalar64) {
+	var p64 scalar64
+	prod := (&[_N + 1]uint64{})[:]
 	mulWord(prod, x64[:], y64[_N-1])
-	copy(z64[:], prod[:_N])
-	z64.reduceOneWord(prod[_N])
+	copy(p64[:], prod[:_N])
+	p64.reduceOneWord(prod[_N])
 	for i := _N - 2; i >= 0; i-- {
-		h := z64.leftShift(0)
-		z64.reduceOneWord(h)
+		h := p64.leftShift(0)
+		p64.reduceOneWord(h)
 		mulWord(prod, x64[:], y64[i])
-		c := add(z64[:], z64[:], prod[:_N])
-		z64.reduceOneWord(prod[_N] + c)
+		c := add(p64[:], p64[:], prod[:_N])
+		p64.reduceOneWord(prod[_N] + c)
 	}
-	z64.modOrder()
-	z64.toScalar(z)
+	*z64 = p64
 }
 
-// IsZero returns true if z=0.
-func (z *Scalar) IsZero() bool { z.Red(); return *z == Scalar{} }
+// Inv calculates z = 1/x mod order.
+func (z *Scalar) Inv(x *Scalar) {
+	var x64 scalar64
+	x64.fromScalar(x)
+
+	var T [16]scalar64
+	T[0] = scalar64{1}
+	for i := 1; i < 16; i++ {
+		coremul(&T[i], &T[i-1], &x64)
+	}
+
+	t := &scalar64{1}
+	for i := 8*len(orderMinusTwo) - 4; i >= 0; i -= 4 {
+		b := (orderMinusTwo[i/8] >> uint(i%8)) & 0xF
+		coremul(t, t, t)
+		coremul(t, t, t)
+		coremul(t, t, t)
+		coremul(t, t, t)
+		if b != 0 {
+			coremul(t, t, &T[b])
+		}
+	}
+	t.modOrder()
+	t.toScalar(z)
```

Comment:
This is nice and constant time!

In the future we might want to replace it with an addition chain.  I found one for you with [addchain](https://github.com/mmcloughlin/addchain):

```
$ ~/go/bin/addchain search '2^446 - 13818066809895115352007386748515426880336692474882178609894547503885 - 2'
addchain: expr: "2^446 - 13818066809895115352007386748515426880336692474882178609894547503885 - 2"
addchain: hex: 3fffffffffffffffffffffffffffffffffffffffffffffffffffffff7cca23e9c44edb49aed63690216cc2728dc58f552378c292ab5844f1
addchain: dec: 181709681073901722637330951972001133588410340171829515070372549795146003961539585716195755291692375963310293709091662304773755859649777
addchain: best: opt(dictionary(hybrid(5,0),continued_fractions(dichotomic)))
_10      = 2*1
_11      = 1 + _10
_100     = 1 + _11
_101     = 1 + _100
_1001    = _100 + _101
_1011    = _10 + _1001
_1101    = _10 + _1011
_1111    = _10 + _1101
_10001   = _10 + _1111
_10011   = _10 + _10001
_10101   = _10 + _10011
_10111   = _10 + _10101
_11001   = _10 + _10111
_11011   = _10 + _11001
_11101   = _10 + _11011
_11111   = _10 + _11101
_111110  = 2*_11111
_1111100 = 2*_111110
i24      = _1111100 << 5 + _1111100
i41      = (i24 << 4 + _111110) << 11 + i24
i73      = (i41 << 4 + _111110) << 26 + i41
i129     = i73 << 55 + i73
x222     = i129 << 110 + _11 + i129
i262     = ((x222 << 6 + _11111) << 7 + _11001) << 6
i279     = ((_10001 + i262) << 8 + _11111) << 6 + _10011
i298     = ((i279 << 5 + _10001) << 8 + _10011) << 4
i312     = ((_1011 + i298) << 6 + _11011) << 5 + _1001
i331     = ((i312 << 6 + _1101) << 6 + _11101) << 5
i343     = ((_10101 + i331) << 5 + _10001) << 4 + _1011
i365     = ((i343 << 5 + _1001) << 7 + 1) << 8
i375     = 2*((_1011 + i365) << 6 + _11001) + 1
i396     = ((i375 << 9 + _10011) << 4 + _1001) << 6
i411     = ((_10001 + i396) << 5 + _10111) << 7 + _1011
i431     = ((i411 << 7 + _1111) << 6 + _10101) << 5
i444     = ((_1001 + i431) << 8 + _11011) << 2 + _11
i464     = ((i444 << 5 + _11) << 7 + _101) << 6
i478     = ((_1001 + i464) << 6 + _10101) << 5 + _1101
i498     = ((i478 << 3 + _11) << 9 + _10001) << 6
return     (_1111 + i498) << 4 + 1
```

Reply:
cool, I included it